pallets/asset-registry/src/lib.rs

```diff
@@ -265,181 +323,170 @@ pub mod pallet {
-
-			ensure!(
-				Self::asset_ids(&bounded_name).is_none(),
-				Error::<T>::AssetAlreadyRegistered
-			);
-
-			let asset_id =
-				Self::register_asset(bounded_name, asset_type, existential_deposit, asset_id, xcm_rate_limit)?;
-
-			if let Some(meta) = metadata {
-				let symbol = Self::to_bounded_name(meta.symbol)?;
-				AssetMetadataMap::<T>::insert(
-					asset_id,
-					AssetMetadata {
-						symbol: symbol.clone(),
-						decimals: meta.decimals,
-					},
-				);
-
-				Self::deposit_event(Event::MetadataSet {
-					asset_id,
-					symbol,
-					decimals: meta.decimals,
-				});
-			}
+			let bounded_name = if let Some(name) = name {
+				let bounded_name = Self::to_bounded_name(name)?;
+				Some(bounded_name)
+			} else {
+				None
+			};
 
-			if let Some(loc) = location {
-				ensure!(asset_id != T::NativeAssetId::get(), Error::<T>::CannotUpdateLocation);
-				ensure!(
-					Self::location_assets(&loc).is_none(),
-					Error::<T>::LocationAlreadyRegistered
-				);
-				AssetLocations::<T>::insert(asset_id, &loc);
-				LocationAssets::<T>::insert(&loc, asset_id);
+			let bounded_symbol = if let Some(symbol) = symbol {
+				Some(Self::to_bounded_name(symbol)?)
+			} else {
+				None
+			};
 
-				Self::deposit_event(Event::LocationSet {
-					asset_id,
-					location: loc,
-				});
-			}
+			let details = AssetDetails::new(
+				bounded_name,
+				asset_type,
+				existential_deposit.unwrap_or(DEFAULT_ED),
+				bounded_symbol,
+				decimals,
+				xcm_rate_limit,
+				is_sufficient,
+			);
 
+			Self::do_register_asset(asset_id, &details, location)?;
 			Ok(())
 		}
 
 		/// Update registered asset.
 		///
-		/// Updates also mapping between name and asset id if provided name is different than currently registered.
+		/// All parameteres are optional and value is not updated if param is `None`.
+		///
+		/// `decimals` - can be update by `UpdateOrigin` only if it wasn't set yet. Only
+		/// `RegistryOrigin` can update `decimals` if it was previously set.
+		///
+		/// `location` - can be updated only by `RegistryOrigin`.
 		///
 		/// Emits `Updated` event when successful.
-
-		// TODO: No tests
 		#[pallet::call_index(1)]
 		#[pallet::weight(<T as Config>::WeightInfo::update())]
 		pub fn update(
 			origin: OriginFor<T>,
 			asset_id: T::AssetId,
-			name: Vec<u8>,
-			asset_type: AssetType<T::AssetId>,
-			existential_deposit: Option<T::Balance>,
-			xcm_rate_limit: Option<T::Balance>,
+			name: Option<Vec<u8>>,
+			asset_type: Option<AssetType<T::AssetId>>,
+			existential_deposit: Option<Balance>,
+			xcm_rate_limit: Option<Balance>,
+			is_sufficient: Option<bool>,
+			symbol: Option<Vec<u8>>,
+			decimals: Option<u8>,
+			location: Option<T::AssetNativeLocation>,
 		) -> DispatchResult {
-			T::RegistryOrigin::ensure_origin(origin)?;
+			let is_registry_origin = match T::RegistryOrigin::ensure_origin(origin.clone()) {
+				Ok(_) => true,
+				Err(_) => {
+					T::UpdateOrigin::ensure_origin(origin)?;
+
+					false
+				}
+			};
```

Comment:
just nitpick really.

i would probably go for something like this. It looks more readable

```rust
let is_registry_origin = T::RegistryOrigin::ensure_origin(origin.clone()).is_ok();
if !is_registry_origin {
	T::UpdateOrigin::ensure_origin(origin)?;
}
```